persistence/javadsl/src/main/scala/com/lightbend/lagom/internal/javadsl/persistence/ReadSideActor.scala

```diff
@@ -3,103 +3,108 @@
+      clazz,
+      globalPrepareTask,
+      eventStreamFactory,
+      processor,
+      mat
+    )
   }
 
-  /**
-   * Start processing from the given offset
-   */
-  case class Start(offset: Offset)
+  case object Start
+
 }
 
 /**
  * Read side actor
  */
 private[lagom] class ReadSideActor[Event <: AggregateEvent[Event]](
-  processorFactory:     () => ReadSideProcessor[Event],
-  eventStreamFactory:   (AggregateEventTag[Event], Offset) => Source[akka.japi.Pair[Event, Offset], NotUsed],
-  clazz:                Class[Event],
-  globalPrepareTask:    ClusterStartupTask,
-  globalPrepareTimeout: FiniteDuration
+  config:             ReadSideConfig,
+  clazz:              Class[Event],
+  globalPrepareTask:  ClusterStartupTask,
+  eventStreamFactory: (AggregateEventTag[Event], Offset) => Source[akka.japi.Pair[Event, Offset], NotUsed],
+  processorFactory:   () => ReadSideProcessor[Event]
 )(implicit mat: Materializer) extends Actor with ActorLogging {
+
   import ReadSideActor._
   import akka.pattern.pipe
-  import akka.pattern.ask
   import context.dispatcher
 
   private var shutdown: Option[KillSwitch] = None
 
+  override def preStart(): Unit = {
+    super.preStart()
+  }
+
   override def postStop: Unit = {
     shutdown.foreach(_.shutdown())
   }
 
   def receive = {
     case EnsureActive(tagName) =>
-
-      val tag = AggregateEventTag.of(clazz, tagName)
-
-      implicit val timeout = Timeout(globalPrepareTimeout)
-
-      globalPrepareTask.askExecute() pipeTo self
-      context become preparing(tag)
-  }
-
-  def preparing(tag: AggregateEventTag[Event]): Receive = {
-
-    case Done =>
-      val handler = processorFactory().buildHandler()
-      handler.prepare(tag).toScala.map(Start) pipeTo self
-      context become active(handler, tag)
-
-    case Status.Failure(e) =>
-      throw e
-
+      implicit val timeout = Timeout(config.globalPrepareTimeout)
+      globalPrepareTask.askExecute().map { _ => Start } pipeTo self
+      context.become(start(tagName))
   }
 
-  def active(handler: ReadSideProcessor.ReadSideHandler[Event], tag: AggregateEventTag[Event]): Receive = {
-
-    case Start(offset) =>
-
-      val (killSwitch, streamDone) = eventStreamFactory(tag, offset).asScala
+  def start(tagName: EntityId): Receive = {
+    case Start =>
+      val tag = new AggregateEventTag(clazz, tagName)
+      val backoffSource =
+        RestartSource.withBackoff(
+          config.minBackoff,
+          config.maxBackoff,
+          config.randomBackoffFactor
+        ) { () =>
+          val handler: ReadSideProcessor.ReadSideHandler[Event] = processorFactory().buildHandler()
+          val futureOffset = handler.prepare(tag).toScala
+          scaladsl.Source.fromFuture(futureOffset).flatMapConcat {
+            offset =>
+              val eventStreamSource = eventStreamFactory(tag, offset).asScala
+              val userlandFlow = handler.handle()
+              eventStreamSource.via(userlandFlow)
+          }
+
+        }
```

Comment:
Builds the `RestartSource`. That's a 3 step process:

- setup the backoff supervision
- create a `Source` from the `Future[Offset]`
- `flatMapConcat` that `Source` into the creation of a `Source` resulting from connecting the `eventStream` and the userland `Flow`